### PR TITLE
For #7134 - Dismiss the current select login prompt request in LoginPicker

### PR DESCRIPTION
**Pull request by ekager**
---

@jonalmeida i couldn't repro the crash myself but I realized I could have been dismissing the wrong request so I'm hoping this is the fix 

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by jonalmeida** [comment]
@ekager I found a consistent set of steps to reproduce the crash I had mentioned on the other PR. I'm going to try and reproduce them with this patch as well. :)

### Steps to reproduce
  1. Go to a site.
  2. Click on the password text field.
  3. Click on a saved password to autofill (click nothing else).
  4. Open Settings from kebab menu.
  5. Go back to the browser.
  6. Crash! :boom:

**Comment by ekager** [comment]
Ah I can repro. I'll update when I figure out what's happening

**Comment by jonalmeida** [comment]
Hmm it still crashes for me, but I think this is something we should fix.

I noticed there is a discrepancy `consumePromptFrom` which looks for the request from the selected tab OR the custom tab. Would that might mean we have a login prompt in a custom tab that we never never dismiss since we don't support logins there?

**Comment by ekager** [comment]
It looks like when you return from settings, we get two simultaneous calls to dismiss the prompt from the `dismissPromptScope` and `sessionPromptScope` which causes the crash. We could try/catch the dismiss, but it seems better to just not try to dismiss at all if we have already tried. What do you think?

**Comment by ekager** [comment]
Oh and to answer your other question - this does work in custom tabs.

**Comment by ekager** [comment]
Okay this is ready for review @Amejia481 . I tried to consolidate the two scopes that dismiss prompts into one and switched to using the `url` instead of the `progress`. I don't think we can use the progress because it would require us to `filter` and I was having a hard time thinking of a case with prompts that we shouldn't dismiss when the page changes. Let me know what you think (and sorry this is on a time crunch!)

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/login/LoginPicker.kt [hunk 1]
```diff
@@ -34,21 +33,13 @@ internal class LoginPicker(
...
 
-    @VisibleForTesting
-    internal var currentRequest: PromptRequest.SelectLoginPrompt? = null
-
     override fun onLoginSelected(login: Login) {
-        (currentRequest as PromptRequest.SelectLoginPrompt).onConfirm(login)
-        currentRequest = null
+        store.consumePromptFrom(sessionId) {
+            if (it is PromptRequest.SelectLoginPrompt) it.onConfirm(login)
+        }
```
Comment (Amejia481) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/login/LoginPicker.kt:42]: Should we include tests for these new behaviour?

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/login/LoginPicker.kt [hunk 2]
```diff
@@ -58,12 +49,9 @@ internal class LoginPicker(
     }
 
     fun dismissCurrentLoginSelect() {
-        if (currentRequest != null) {
-            store.consumePromptFrom(sessionId) {
-                (it as PromptRequest.SelectLoginPrompt).onDismiss()
-            }
+        store.consumePromptFrom(sessionId) {
+            if (it is PromptRequest.SelectLoginPrompt) it.onDismiss()
```
Comment (Amejia481) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/login/LoginPicker.kt:53]: Should we include tests for these new behaviour?

**Comment by Amejia481** [comment]
There are some test failing on `build-feature-prompts` let's take a look :)

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/login/LoginPicker.kt [hunk 3]
```diff
@@ -57,13 +48,15 @@ internal class LoginPicker(
...
-            store.consumePromptFrom(sessionId) {
-                (it as PromptRequest.SelectLoginPrompt).onDismiss()
+    @Suppress("TooGenericExceptionCaught")
+    fun dismissCurrentLoginSelect(promptRequest: PromptRequest.SelectLoginPrompt? = null) {
+        try {
+            promptRequest?.let { it.onDismiss() } ?: store.consumePromptFrom(sessionId) {
+                if (it is PromptRequest.SelectLoginPrompt) it.onDismiss()
             }
+        } catch (e: RuntimeException) {
+            // This prompt has already been dismissed
```
Comment (Amejia481) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/login/LoginPicker.kt:58]: nit: let's log the exception

**Comment by ekager** [comment]
bors r=Amejia481

**Pull request merged**